pkg/utils/utils.go

```diff
@@ -1034,3 +1036,171 @@ func CreatePluginClientSet() (*plugin_clientset.Clientset, error) {
+	if err == nil {
+		if csiVersion, err = GetCSIVersionFromImage(csiStatefulset.Spec.Template.Spec.Containers); err != nil {
+			// Image not found.
+			return "", err
+		}
+		if csiVersion == constants.CsiDevVersion {
+			// Using unreleased version, defaulting to 1.0.2 since vSphere CSI driver is installed as a
+			// StatefulSet on `kube-system` namespace for versions greater than v1.0.2 and lesser than v2.0.0
+			return constants.CsiMinVersion, nil
+		}
+		return csiVersion, nil
+	}
+
+	// vSphere CSI controller not installed.
+	return csiVersion, errors.Errorf("vSphere CSI controller, %s, is required by velero-plugin-for-vsphere. Please make sure the vSphere CSI controller is installed in the cluster", constants.VSphereCSIController)
+}
+
+func GetCSIClusterType(kubeClient kubernetes.Interface) (constants.ClusterFlavor, error) {
+	csiClusterType := constants.Unknown
+	// For CSI >=2.3.0, the CSI Deployment is in vmware-system-csi namespace.
+	csiDeployment, err := kubeClient.AppsV1().Deployments(constants.VSphereCSIControllerNamespace).Get(context.TODO(), constants.VSphereCSIController, metav1.GetOptions{})
+	if err == nil {
+		if csiClusterType, err = GetCSIClusterTypeFromEnv(csiDeployment.Spec.Template.Spec.Containers); err != nil {
+			// Image not found.
+			return constants.Unknown, err
+		}
+		return csiClusterType, nil
+	}
+	// Search in kube-system next.
+	csiDeployment, err = kubeClient.AppsV1().Deployments(constants.KubeSystemNamespace).Get(context.TODO(), constants.VSphereCSIController, metav1.GetOptions{})
+	if err == nil {
+		if csiClusterType, err = GetCSIClusterTypeFromEnv(csiDeployment.Spec.Template.Spec.Containers); err != nil {
+			// Image not found.
+			return constants.Unknown, err
+		}
+		return csiClusterType, nil
+	}
+	// CSI driver is deployed as StatefulSet for v1.0.2 and v1.0.3.
+	csiStatefulset, err := kubeClient.AppsV1().StatefulSets(constants.KubeSystemNamespace).Get(context.TODO(), constants.VSphereCSIController, metav1.GetOptions{})
+	if err == nil {
+		if csiClusterType, err = GetCSIClusterTypeFromEnv(csiStatefulset.Spec.Template.Spec.Containers); err != nil {
+			// Image not found.
+			return constants.Unknown, err
+		}
+		return csiClusterType, nil
+	}
+	// vSphere CSI controller not installed.
+	return csiClusterType, errors.Errorf("vSphere CSI controller, %s, is required by velero-plugin-for-vsphere. Please make sure the vSphere CSI controller is installed in the cluster", constants.VSphereCSIController)
+}
+
+func GetCSIVersionFromImage(containers []k8sv1.Container) (string, error) {
+	var csiDriverVersion string
+	csiDriverVersion = GetVersionFromImage(containers, "cloud-provider-vsphere/csi/release/driver")
+	if csiDriverVersion == "" {
+		csiDriverVersion = GetVersionFromImage(containers, "cloud-provider-vsphere/csi/ci/driver")
+		if csiDriverVersion == "" {
+			csiDriverVersion = GetVersionFromImage(containers, "vsphere-csi-block-driver")
```

Comment:
why do we need to check `vsphere-csi-block-driver`? where does it come from?

Reply:
it is not a good idea to add code dependency on the name of repo. What does the image name without repo look like? Can we rely on that? Basically, if CSI team decide to move their repo to one another, we need to make a code change. That is not ideal.

Reply:
This is only to test on the unreleased version. If you notice the if structure this is absolutely the last resort to determine the version, once 2.3.0 is released on the correct repo we will never hit the condition. If they change the repo(unlikely) the worst thing that could happen is our internal testing will fail and won't affect real workloads. The current qe test repo for csi is like `csi-driver-and-syncer-docker-local.artifactory.eng.vmware.com/vsphere-csi-block-driver:v2.1.0-rc.1-664-g2d5a5df`.

Reply:
For testing unreleased 2.3.0 images like this, our test code can definitely rename the image to anything. That is way to test it rather than hacd-coding the repo in production code.

Reply:
I've requested the QE several times, but those changes are very slow to come by, I cannot block my changes waiting for them, I need some way to test my changes, this is the only way I see. This is harmless, once they make their changes I'll probably delete this before our release. Like I've said, this block will never get entered on official release images.

Reply:
Ok made those changes, had to re-tag and push again.